tests/test_weighted_random_sampler.py

```diff
@@ -7,55 +7,104 @@
 pytestmark = [pytest.mark.memcheck]
 
 
-@pytest.mark.parametrize("seed", [2022])
-@pytest.mark.parametrize("size, num_samples", [(500, 1500), (1000, 3000)])
+@pytest.mark.parametrize("seed", [2024])
+@pytest.mark.parametrize("size, num_samples", [(500, 1500), (1000, 3000), (1000, 10000)])
 @pytest.mark.parametrize("precision", [np.float32, np.float64])
-def test_random_sampler(seed, size, num_samples, precision):
+def test_segmented_random_sampler(seed, size, num_samples, precision):
     rng = np.random.default_rng(seed)
 
     # Get a random probability vector.
-    probs = rng.exponential(scale=1 / 250, size=size)
-    prob_total = np.sum(probs)
-    probs /= prob_total
-    # Sort the probabilities to easily visualize
-    probs = np.sort(probs)[::-1]
+    probs = rng.exponential(scale=1 / 250, size=(num_samples, size))
 
-    klass = custom_ops.WeightedRandomSampler_f32
+    klass = custom_ops.SegmentedWeightedRandomSampler_f32
     if precision == np.float64:
-        klass = custom_ops.WeightedRandomSampler_f64
+        klass = custom_ops.SegmentedWeightedRandomSampler_f64
 
     x = np.arange(size)
-    ref_selection = rng.choice(x, size=num_samples, p=probs)
-    sampler = klass(size, seed)
+    ref_selection = np.array([rng.choice(x, size=1, p=batch / np.sum(batch)) for batch in probs]).reshape(-1)
+    sampler = klass(size, num_samples, seed)
 
-    test_selection = sampler.sample(num_samples, probs)
+    test_selection = sampler.sample(probs.tolist())
     assert len(test_selection) == num_samples
 
     ks, pv = ks_2samp(ref_selection, test_selection)
     assert ks < 0.05, (pv, ks)
```

Comment:
non-blocking, but would be good to strengthen this test in two ways:
* increase difficulty of instances -- make weights in each of the random instances much further from uniform (i.e. sample each row of probs from dirichlet, with high concentration parameter)
* increase sensitivity of distribution comparison -- maybe by comparing to probs directly, rather than comparing to a single random sample

observation: the test still passes ~95% of the time if the instances in test_selection are shuffled:

```python
def test_segmented_random_sampler(seed, size, num_samples):
    rng = np.random.default_rng(seed)

    # Get a random probability vector.
    probs = rng.exponential(scale=1 / 250, size=(num_samples, size))

    x = np.arange(size)
    ref_selection = np.array([rng.choice(x, size=1, p=batch / np.sum(batch)) for batch in probs]).reshape(-1)
    test_selection = np.array([rng.choice(x, size=1, p=batch / np.sum(batch)) for batch in probs]).reshape(-1)
    # np.random.shuffle(test_selection)  # still passes 95% of the time
    assert len(test_selection) == num_samples

    ks, pv = ks_2samp(ref_selection, test_selection)
    assert ks < 0.05, (pv, ks)

    return probs, ref_selection, test_selection, ks, pv
```

Reply:
good catch, +1 to strengthening this test

Reply:
Strengthened the test here: https://github.com/proteneer/timemachine/pull/1235/commits/eefa7867b083747f7aaf76f09aed90fd700b8301

Uses the dirchlet distribution for each sample as well as compares the percentages rather than the ks test

Reply:
Realized the test was way too loose and fixed it up in https://github.com/proteneer/timemachine/pull/1235/commits/1da7ca7cda67b839f19906686df8e31e12bd5320